pkg/vpcmodel/connectivityAbstraction.go

```diff
@@ -0,0 +1,159 @@
+/*
+Copyright 2023- IBM Inc. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package vpcmodel
+
+import (
+	"fmt"
+	"slices"
+)
+
+// connectivity abstraction of a nodeSet:
+// consider nodeSet NS[N0, N1, N2, N3...]
+// we assume that for every node N,  if we have the connection  N->N0
+// then we also have the connections N->N1, N->N2, N->N3 ...
+// we will call it the abstraction assumption
+// so we replace all these connections with one connection: N->NS.
+
+// the abstraction steps are:
+// 1. splitting the connectivity to for groups:
+//      otherToOther:     connections of   <node not in the nodeSet>  ->  <node not in the nodeSet>
+//      nodeSetToNodeSet: connections of   <node     in the nodeSet>  ->  <node     in the nodeSet>
+//      otherFromNodeSet: connections of   <node     in the nodeSet>  ->  <node not in the nodeSet>
+//      otherToNodeSet:   connections of   <node not in the nodeSet>  ->  <node     in the nodeSet>
+// 2. for the last three groups, we check the abstraction assumption holds
+// 3. we do the abstraction ( for now, even if the abstraction assumption does not hold):
+// the connectivity of N->NS is union of all of N->N1, N->N2, N->N3 ...
+// todo: what to do if the abstraction assumption does not hold?
+
+func nodeSetConnectivityAbstraction(nodesConn GeneralConnectivityMap, nodeSet NodeSet) GeneralConnectivityMap {
+	otherToOther, nodeSetToNodeSet, otherFromNodeSet, otherToNodeSet := splitConnectivityByNodeSet(nodesConn, nodeSet)
+	checkConnectivityAbstractionValidity(otherFromNodeSet, nodeSet, false)
+	checkConnectivityAbstractionValidity(otherToNodeSet, nodeSet, true)
+	checkConnectivityAbstractionValidity(nodeSetToNodeSet, nodeSet, true)
+	return mergeConnectivityWithNodeSetAbstraction(otherToOther, nodeSetToNodeSet, otherFromNodeSet, otherToNodeSet, nodeSet)
+}
+
+// splitConnectivityByNodeSet() split the connectivity to the four groups
+// each group is kept as GeneralConnectivityMap.
+// usually, GeneralConnectivityMap is a map form src to dst.
+// however, the third group is hold as a map from dst to src (and therefore called otherFromNodeSet and not nodeSetToOther)
+// see the reason on mergeConnectivityWithNodeSetAbstraction()
+
+func splitConnectivityByNodeSet(nodesConn GeneralConnectivityMap, nodeSet NodeSet) (
+	otherToOther, nodeSetToNodeSet, otherFromNodeSet, otherToNodeSet GeneralConnectivityMap) {
+	otherToOther = GeneralConnectivityMap{}
+	nodeSetToNodeSet = GeneralConnectivityMap{}
+	otherFromNodeSet = GeneralConnectivityMap{}
+	otherToNodeSet = GeneralConnectivityMap{}
+	for src, nodeConns := range nodesConn {
+		for dst, conns := range nodeConns {
+			srcNode, srcIsNode := src.(Node)
+			dstNode, dstIsNode := dst.(Node)
+			srcInSet := srcIsNode && slices.Contains(nodeSet.Nodes(), srcNode)
+			dstInSet := dstIsNode && slices.Contains(nodeSet.Nodes(), dstNode)
+			switch {
```

Comment:
Why aren't you doing the abstraction here - namely, why aren't you updating the connection with the `NodeSet` where relevant and are using the original `Nodes` instead?

Reply:
not sure I understand the question,
I think that it is a good practice to split this to three steps
1. partitioning
2. checking
3. merging
later we will probably want to do checking wo merging, or merging if checking is OK